File: GK_Updater_Module/ProgressCounter.vb
Public Class ProgressCounter

#Region "Fields"

    Private _currentTick As Integer
    Private _progBytesMoved As Integer
    Private _progTotalBytes As Integer
    Private _speedBytesMoved As Integer
    Private _speedThroughput As Double
    Private _startTick As Integer

#End Region

#Region "Constructors"

    Sub New()
        _progBytesMoved = 0
        _progTotalBytes = 0
        _speedBytesMoved = 0
        _currentTick = 0
        _startTick = 0
        _speedThroughput = 0
    End Sub

#End Region

#Region "Properties"

    Public Property BytesMoved As Integer
        Get
            Return _progBytesMoved
        End Get
        Set(value As Integer)
            _speedBytesMoved += value
            _progBytesMoved += value
        End Set
    End Property

    Public Property BytesToTransfer As Integer
        Get
            Return _progTotalBytes
        End Get
        Set(value As Integer)
            _progTotalBytes = value
        End Set
    End Property

    Public ReadOnly Property Percent As Integer
        Get
            If _progTotalBytes > 0 Then
                Return CInt((_progBytesMoved / _progTotalBytes) * 100)
            Else
                Return 0
            End If
        End Get
    End Property
    Public ReadOnly Property Throughput As Double
        Get
            Return _speedThroughput
        End Get
    End Property

#End Region

#Region "Methods"

    Public Sub ResetProgress()
        _progBytesMoved = 0
    End Sub

    Public Sub Tick()
        _currentTick = Environment.TickCount
        If _startTick > 0 Then
            If _speedBytesMoved > 0 Then
                Dim elapTime = _currentTick - _startTick
                _speedThroughput = Math.Round((_speedBytesMoved / elapTime) / 1000, 2)
            End If
        Else
            _startTick = _currentTick
        End If
    End Sub

#End Region

End Class

File: AssetManager/Tools/ProgressCounter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AssetManager
{
    public class ProgressCounter
    {

        #region "Fields"

        private int _currentTick;
        private int _progBytesMoved;
        private int _progTotalBytes;
        private int _speedBytesMoved;
        private double _speedThroughput;

        private int _startTick;
        #endregion

        #region "Constructors"

        public ProgressCounter()
        {
            _progBytesMoved = 0;
            _progTotalBytes = 0;
            _speedBytesMoved = 0;
            _currentTick = 0;
            _startTick = 0;
            _speedThroughput = 0;
        }

        #endregion

        #region "Properties"

        public int BytesMoved
        {
            get { return _progBytesMoved; }
            set
            {
                _speedBytesMoved += value;
                _progBytesMoved += value;
            }
        }

        public int BytesToTransfer
        {
            get { return _progTotalBytes; }
            set { _progTotalBytes = value; }
        }

        public int Percent
        {
            get
            {
                if (_progTotalBytes > 0)
                {
                    return (int)Math.Round((((float)_progBytesMoved / (float)_progTotalBytes) * 100), 0);
                }
                else
                {
                    return 0;
                }
            }
        }

        public double Throughput
        {
            get { return _speedThroughput; }
        }

        #endregion

        #region "Methods"

        public void ResetProgress()
        {
            _progBytesMoved = 0;
        }

        public void Tick()
        {
            _currentTick = Environment.TickCount;
            if (_startTick > 0)
            {
                if (_speedBytesMoved > 0)
                {
                    double elapTime = _currentTick - _startTick;
                    _speedThroughput = Math.Round((_speedBytesMoved / elapTime) / 1000, 2);
                }
            }
            else
            {
                _startTick = _currentTick;
            }
        }

        #endregion

    }
}